### PR TITLE
Basic support tpch

### PR DESCRIPTION
**Pull request by ouyuanning**
**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [x] Code Refactoring

**Which issue(s) this PR fixes:**

issue #2372 

**What this PR does / why we need it:**

basic support for tpc-h query.

**Special notes for your reviewer:**

1、tpc-h test in plan2/tpch_test.go
2、other test case in plan2/build_test.go
3、Monday meeting's todo will be done next time.

this PR only let building tpc-h query without errors.
then other guys can do their work now.

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

File: pkg/sql/plan2/build_subquery.go [hunk 1]
```diff
@@ -63,25 +47,40 @@ func buildSubQuery(subquery *tree.Subquery, ctx CompilerContext, query *Query, s
...
 
-	//把子查询的节点放到前面去
-	query.Steps = []int32{int32(nowLength - 1)}
-	query.Nodes = append(query.Nodes[nowLength:], query.Nodes[:nowLength]...)
+	//move subquery node to top
+	for i := len(query.Nodes) - 1; i >= 0; i-- {
+		if query.Nodes[i].NodeId == nodeId {
+			query.Nodes = append(query.Nodes[i+1:], query.Nodes[:i+1]...)
+			break
+		}
```
Comment (aunjgr) [inline pkg/sql/plan2/build_subquery.go:61]: Moving nodes around is error-prone. And no need to do that.

**Review by yingfeng — CHANGES_REQUESTED** [review]
Merge upstream to pass the fmt CI check

File: pkg/sql/plan2/build_subquery.go (hunk 1 above)
Comment (ouyuanning) [reply pkg/sql/plan2/build_subquery.go:61]: sometime when you will append something to PreNode,  If the pre node is subquery, that will be error:
eg: select a from tbl where b>(subquery) limit 10
I will improve the code after test case cover more corner

**Review by qingxinhome — APPROVED** [review]

File: pkg/sql/plan2/build_expr.go [hunk 2]
```diff
@@ -89,56 +88,168 @@ func buildExpr(stmt tree.Expr, ctx CompilerContext, query *Query, selectCtx *Sel
...
+				Args: []*plan.Expr{expr, typeExpr},
+			},
+		},
+		Typ: &plan.Type{
+			Id: plan.Type_ANY,
+		},
+	}, nil
+}
+
+func biuldCase(astExpr *tree.CaseExpr, ctx CompilerContext, query *Query, selectCtx *SelectContext) (*plan.Expr, error) {
```
Comment (yingfeng) [inline pkg/sql/plan2/build_expr.go:173]: typo here

File: pkg/sql/plan2/build_expr.go [hunk 3]
```diff
@@ -285,11 +422,11 @@ func buildUnaryExpr(expr *tree.UnaryExpr, ctx CompilerContext, query *Query, sel
 	case tree.UNARY_PLUS:
 		return getFunctionExprByNameAndExprs("UNARY_PLUS", []tree.Expr{expr.Expr}, ctx, query, selectCtx)
 	case tree.UNARY_TILDE:
-		return nil, errors.New(errno.SyntaxErrororAccessRuleViolation, fmt.Sprintf("'%v' is not support now", expr))
+		return nil, errors.New(errno.SyntaxErrororAccessRuleViolation, fmt.Sprintf("'%v' is not support now3", expr))
```
Comment (yingfeng) [inline pkg/sql/plan2/build_expr.go:425]: what does now3,now4,now5 mean?

File: pkg/sql/plan2/build_expr.go [hunk 4]
```diff
@@ -274,8 +375,44 @@ func buildComparisonExpr(expr *tree.ComparisonExpr, ctx CompilerContext, query *
 		return getFunctionExprByNameAndExprs("<>", []tree.Expr{expr.Left, expr.Right}, ctx, query, selectCtx)
```
[the remark this answers is not in the document]
Comment (ouyuanning) [reply pkg/sql/plan2/build_expr.go:375]: good one。may be we can use enum in function.go later

File: pkg/sql/plan2/build_update.go [hunk 5]
```diff
@@ -84,54 +84,55 @@ func buildUpdate(stmt *tree.Update, ctx CompilerContext, query *Query) error {
...
 
+		columns = append(columns, column)
 		values = append(values, value)
 	}
 
-	if len(columns) == 0 {
-		return errors.New(errno.CaseNotFound, "no column will be update")
-	}
-
 	node := &plan.Node{
```
[the remark this answers is not in the document]
Comment (ouyuanning) [reply pkg/sql/plan2/build_update.go:125]: Get it。I will add TableDef/ObjectRef  to  Update/Delete node in next PR.

File: pkg/sql/plan2/build_delete.go [hunk 6]
```diff
@@ -35,7 +35,6 @@ func buildDelete(stmt *tree.Delete, ctx CompilerContext, query *Query) error {
 		Limit:   stmt.Limit,
 	}
 	selectCtx := &SelectContext{
-		tableAlias:  make(map[string]string),
 		columnAlias: make(map[string]*plan.Expr),
 	}
 	return buildSelect(selectStmt, ctx, query, selectCtx)
```
[the remark this answers is not in the document]
Comment (ouyuanning) [reply pkg/sql/plan2/build_delete.go:40]: Get it。I will add TableDef/ObjectRef to Update/Delete node in next PR.

File: pkg/sql/plan2/build_expr.go (hunk 3 above)
Comment (ouyuanning) [reply pkg/sql/plan2/build_expr.go]: that's debug code.
fixed now

File: pkg/sql/plan2/build_expr.go (hunk 2 above)
Comment (ouyuanning) [reply pkg/sql/plan2/build_expr.go]: thanks, fixed now

**Review by yingfeng — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]
Let's merge it and fix later, not to block development.

**Pull request merged**